# Exercise 1: Jupyter Notebooks and Python

* Work through the short tutorial below to get you up to speed in python and jupyter notebooks. 

## Tutorial on Jupyter and Python 

### Jupyter Notebooks

* Cells can be <b>evaluated</b> by selecting them and then hitting <code>shift + enter</code>
* New cells can be <b>inserted</b> using the Insert option in the menu
* Cells can be <b>edited</b> using the Edit option in the menu

In [2]:
# This is a cell. You can execute it by selecting it using your mouse (indicated by a border) and then press shift+enter. 
print('You have executed this cell.')

You have executed this cell.



### Python

* Python is a strongly, dynamically <b>typed</b> language
* Code blocks in python are defined using <b>indents</b> (tabs) instead of <code>}</code> or <code>{</code>
* You do not have to use <code>;</code>
* <code>)</code> and <code>(</code> can be omitted in many cases 
* <b>Lists, dictionaries and numpy arrays</b> will be our main data structures

Try to <b>learn</b> the syntax from <b>coding</b>, and referring to other code samples. It is quite simple. We will cover
only the most important things here.

### Conditional Statements

In python there is an <code>if</code> statement (the indent is important!) following the same semantical rules as in other languages:

```python
if condition:
    statement
```

In [3]:
if 10 > 5:
    print('10 is greater than 1.')

10 is greater than 1.


There is also an else and an else-if version of the statement. Logical expression are negated using the <code>not</code>
statement, they can be concatenated using the <code>and</code> and <code>or</code> operators:<br><br>
``` python
if condition1:
    statement1
elif condition2:
    statement2
else condition3:
    statement3
```


In [4]:
x = 3.001

if x < 0:
    print('x is in (-inf, 0)')
elif x >= 0 and x <= 3:
    print('x is in [0;3]') 
else:
    print('x is in (3;inf)')

x is in (3;inf)


### Python Datastructures

#### Lists
Lists and dictionaries and tuples are the most important python data structures. A list is defined using the <code>[]</code> syntax.

In [5]:
list1 = [] # This defines a list
list1.append('item1') # add a string to the list
list1.append(2) # add an integer to the list, lists are not restricted to one data-type

list1.append(list1)
print(list1)

['item1', 2, [...]]


In [6]:
list2 = ['item1', 2] # direct initialization of a list
print(list2)

['item1', 2]


In [7]:
list3 = ['item1', 2, 'item2', 3]
list3.remove('item1') # items can be removed
print(list3)

[2, 'item2', 3]


We can check if an item is inside the list by using the <code>in</code> operator:

In [8]:
list4 = ['item1', 2, 'item2', 3]
print(2 in list4)
print('item1' not in list4)

True
False


You can retrieve the number of items in a list using the <code>len</code> operator:

In [9]:
list1 = [1,2,3,4,5,6]
len(list1)

6

### Dictionaries
Dictionaries are an <b>associative</b> data structure (such as a HashMap) combining a key and a value:

In [10]:
fvecs_by_class = dict() # initialize a dictionary
fvecs_by_class['cat'] = [[1, 2, 3 ,4]] # key is a string, value is a list, containing a single element (another list)
fvecs_by_class['dog'] = [[5, 6, 1, 0]]

print(fvecs_by_class)

fvecs_by_class['cat'].append('bad fvec')

print(fvecs_by_class)

{'cat': [[1, 2, 3, 4]], 'dog': [[5, 6, 1, 0]]}
{'cat': [[1, 2, 3, 4], 'bad fvec'], 'dog': [[5, 6, 1, 0]]}


In [11]:
fvecs_by_class['cat'].append('a bad feature vector') # we add another value to the string which certainly is not a fvec!!
print(fvecs_by_class)

{'cat': [[1, 2, 3, 4], 'bad fvec', 'a bad feature vector'], 'dog': [[5, 6, 1, 0]]}


### Tuples

Tuples are another important data structure we use in python. However, we usually do not use them to represent vectors (we use a library for this called numpy). Nevertheless, tuples can be very useful in many applications. 

They are exactly what you expect them to be, the behave like tuples in math. Remember our definition of tuples using the carthesian product:

$A \times B = \{ (a,b) \;|\; a \in A, b \in B\}$

Tuples are defined in python using the <code>()</code> syntax. As python is not strongly typed, we do not have to define the type of each dimension (as opposed to mathematics).

In [12]:
Colors = ['red', 'green', 'blue', 'yellow']
tuple1 = (1, 3, Colors[0]) # once defined tuples have a fixed numer of elements (as opposed to a list)

print(tuple1) # tuples are accessed using indexing

(1, 3, 'red')


So from a mathematical perspective **tuple1** above but be an element of the set
$\mathbb{R} \times \mathbb{R} \times Colors$. Remember in tuples:
- Ordering matters
- Elements can occur in a tuple multiple times

In [13]:
tuple1 = (1,3,4)
tuple2 = (1,4,3)
print(tuple1 == tuple2) # Ordering matters

tuple3 = (1,1,1) # Elements can occur multiple times
print(tuple3)

False
(1, 1, 1)


### For-Loops

For loops are everywhere in ML and signal processing. The syntax is a bit different from what we are used to in Java and C.
For loops are based on the idea of <b>iterators</b> in python (this is similar to foreach in other languages). We therefore do not define a start-value, end condition and iteration statement like in other languages we simply iterate over a sequence.

The syntax is:
<code>
for variable in iterable-structure:
    statement
</code>

We generally use the <code>range(*start*, *stop*, *step*)</code> function in combination with for loops. 


In [14]:
for value in range(1,10,2):
    print(value)

1
3
5
7
9


In [15]:
list1 = ['one', 'two', 'three']
for item in list1:
    print(item)

one
two
three


### Functions
Functions in python are defined using the <b>def</b> keyword. Be aware that there are no "{" or "}". Indents are used to define the program structure. The general syntax is:
<code>
    **def** func(arg1, arg2):
        # some code here
        **return** value # if you want to return a value
</code>

In [16]:
def add_vectors(tuple1, tuple2): # implements a standard vector addition using tuples as input
    
    # We have to check that:
    #    1. the tuples have the same dimension 
    #    2. the tuples have the same type
    if len(tuple1) != len(tuple2):
        raise Exception('Vectors do not have the same dimension')
    
    # If we need variable dimensional tuples (like k-dimensional) we use lists instead.
    # They are almost the same thing. 
    tuple3 = []
    
    # Loop through all dimensions and perform element-wise addition.
    #
    for i in range(len(tuple1)): 
        tuple3.append(tuple1[i] + tuple2[i]) # Append to the end of the list
    
    return tuple(tuple3) # Convert to tuple

In [17]:
vec1 = (1,2,3)
vec2 = (3,4,5)
add_vectors(vec1, vec2)

(4, 6, 8)

## Numpy

Notice: Parts of this tutorial are from the excellent tutorial found at https://cs231n.github.io/python-numpy-tutorial/#numpy.


Numpy is the core library for scientific computing in Python. It provides a high-performance multidimensional array object, and tools for working with these arrays. If you are already familiar with MATLAB, you might find this tutorial useful to get started with Numpy.


### Arrays
A numpy array is a grid of values, all of the same type, and is indexed by a tuple of nonnegative integers. The number of dimensions is the rank of the array; the shape of an array is a tuple of integers giving the size of the array along each dimension.

We can initialize numpy arrays from nested Python lists, and access elements using square brackets. The first index into the array corresponds to the row-index, the second index to the column-index (such as when we use matrices). The third dimension is the color channel.

<center>
<img src='https://github.com/shegenbart/Jupyter-Exercises/blob/main/Data-Supervised-Unsupervised-RL/figures/numpy_tensor.png?raw=1' width=600/>
</center>


In [18]:
import numpy as np

a = np.array([1, 2, 3])   # Create a rank 1 array
print(type(a))            # Prints "<class 'numpy.ndarray'>"
print(a.shape)            # Prints "(3,)"
print(a[0], a[1], a[2])   # Prints "1 2 3"
a[0] = 5                  # Change an element of the array
print(a)                  # Prints "[5, 2, 3]"

b = np.array([[1,2,3],[4,5,6]])    # Create a rank 2 array
print(b.shape)                     # Prints "(2, 3)"
print(b[0, 0], b[0, 1], b[1, 0])   # Prints "1 2 4"

<class 'numpy.ndarray'>
(3,)
1 2 3
[5 2 3]
(2, 3)
1 2 4


Numpy also provides many functions to create arrays:

In [19]:
import numpy as np

a = np.zeros((2,2))   # Create an array of all zeros
print(a)              # Prints "[[ 0.  0.]
                      #          [ 0.  0.]]"

b = np.ones((1,2))    # Create an array of all ones
print(b)              # Prints "[[ 1.  1.]]"

c = np.full((2,2), 7)  # Create a constant array
print(c)               # Prints "[[ 7.  7.]
                       #          [ 7.  7.]]"

d = np.eye(2)         # Create a 2x2 identity matrix
print(d)              # Prints "[[ 1.  0.]
                      #          [ 0.  1.]]"

e = np.random.random((2,2))  # Create an array filled with random values
print(e)                     # Might print "[[ 0.91940167  0.08143941]
                             #               [ 0.68744134  0.87236687]]"

[[0. 0.]
 [0. 0.]]
[[1. 1.]]
[[7 7]
 [7 7]]
[[1. 0.]
 [0. 1.]]
[[0.35363478 0.06737176]
 [0.99376927 0.90660665]]


### Array indexing
Numpy offers several ways to index into arrays.

Slicing: Similar to Python lists, numpy arrays can be sliced. Since arrays may be multidimensional, you must specify a slice for each dimension of the array:

In [20]:
import numpy as np

# Create the following rank 2 array with shape (3, 4)
# [[ 1  2  3  4]
#  [ 5  6  7  8]
#  [ 9 10 11 12]]
a = np.array([[1,2,3,4], [5,6,7,8], [9,10,11,12]])

# Use slicing to pull out the subarray consisting of the first 2 rows
# and columns 1 and 2; b is the following array of shape (2, 2):
# [[2 3]
#  [6 7]]
b = a[:2, 1:3]

# A slice of an array is a view into the same data, so modifying it
# will modify the original array.
print(a[0, 1])   # Prints "2"
b[0, 0] = 77     # b[0, 0] is the same piece of data as a[0, 1]
print(a[0, 1])   # Prints "77"

2
77


You can also mix integer indexing with slice indexing. However, doing so will yield an array of lower rank than the original array. Note that this is quite different from the way that MATLAB handles array slicing:

In [21]:
import numpy as np

# Create the following rank 2 array with shape (3, 4)
# [[ 1  2  3  4]
#  [ 5  6  7  8]
#  [ 9 10 11 12]]
a = np.array([[1,2,3,4], [5,6,7,8], [9,10,11,12]])

# Two ways of accessing the data in the middle row of the array.
# Mixing integer indexing with slices yields an array of lower rank,
# while using only slices yields an array of the same rank as the
# original array:
row_r1 = a[1, :]    # Rank 1 view of the second row of a
row_r2 = a[1:2, :]  # Rank 2 view of the second row of a
print(row_r1, row_r1.shape)  # Prints "[5 6 7 8] (4,)"
print(row_r2, row_r2.shape)  # Prints "[[5 6 7 8]] (1, 4)"

# We can make the same distinction when accessing columns of an array:
col_r1 = a[:, 1]
col_r2 = a[:, 1:2]
print(col_r1, col_r1.shape)  # Prints "[ 2  6 10] (3,)"
print(col_r2, col_r2.shape)  # Prints "[[ 2]
                             #          [ 6]
                             #          [10]] (3, 1)"

[5 6 7 8] (4,)
[[5 6 7 8]] (1, 4)
[ 2  6 10] (3,)
[[ 2]
 [ 6]
 [10]] (3, 1)


One useful trick with integer array indexing is selecting or mutating one element from each row of a matrix:

In [22]:
import numpy as np

# Create a new array from which we will select elements
a = np.array([[1,2,3], [4,5,6], [7,8,9], [10, 11, 12]])

print(a)  # prints "array([[ 1,  2,  3],
          #                [ 4,  5,  6],
          #                [ 7,  8,  9],
          #                [10, 11, 12]])"

# Create an array of indices
b = np.array([0, 2, 0, 1])

# Select one element from each row of a using the indices in b
print(a[np.arange(4), b])  # Prints "[ 1  6  7 11]"

# Mutate one element from each row of a using the indices in b
a[np.arange(4), b] += 10

print(a)  # prints "array([[11,  2,  3],
          #                [ 4,  5, 16],
          #                [17,  8,  9],
          #                [10, 21, 12]])

[[ 1  2  3]
 [ 4  5  6]
 [ 7  8  9]
 [10 11 12]]
[ 1  6  7 11]
[[11  2  3]
 [ 4  5 16]
 [17  8  9]
 [10 21 12]]


**Boolean array indexing:** Boolean array indexing lets you pick out arbitrary elements of an array. Frequently this type of indexing is used to select the elements of an array that satisfy some condition. Here is an example:

In [23]:
import numpy as np

a = np.array([[1,2], [3, 4], [5, 6]])

bool_idx = (a > 2)   # Find the elements of a that are bigger than 2;
                     # this returns a numpy array of Booleans of the same
                     # shape as a, where each slot of bool_idx tells
                     # whether that element of a is > 2.

print(bool_idx)      # Prints "[[False False]
                     #          [ True  True]
                     #          [ True  True]]"

# We use boolean array indexing to construct a rank 1 array
# consisting of the elements of a corresponding to the True values
# of bool_idx
print(a[bool_idx])  # Prints "[3 4 5 6]"

# We can do all of the above in a single concise statement:
print(a[a > 2])     # Prints "[3 4 5 6]"

[[False False]
 [ True  True]
 [ True  True]]
[3 4 5 6]
[3 4 5 6]


### Array Math

Basic mathematical functions operate elementwise on arrays, and are available both as operator overloads and as functions in the numpy module:

In [24]:
import numpy as np

x = np.array([[1,2],[3,4]], dtype=np.float64)
y = np.array([[5,6],[7,8]], dtype=np.float64)

# Elementwise sum; both produce the array
# [[ 6.0  8.0]
#  [10.0 12.0]]
print(x + y)
print(np.add(x, y))

# Elementwise difference; both produce the array
# [[-4.0 -4.0]
#  [-4.0 -4.0]]
print(x - y)
print(np.subtract(x, y))

# Elementwise product; both produce the array
# [[ 5.0 12.0]
#  [21.0 32.0]]
print(x * y)
print(np.multiply(x, y))

# Elementwise division; both produce the array
# [[ 0.2         0.33333333]
#  [ 0.42857143  0.5       ]]
print(x / y)
print(np.divide(x, y))

# Elementwise square root; produces the array
# [[ 1.          1.41421356]
#  [ 1.73205081  2.        ]]
print(np.sqrt(x))

[[ 6.  8.]
 [10. 12.]]
[[ 6.  8.]
 [10. 12.]]
[[-4. -4.]
 [-4. -4.]]
[[-4. -4.]
 [-4. -4.]]
[[ 5. 12.]
 [21. 32.]]
[[ 5. 12.]
 [21. 32.]]
[[0.2        0.33333333]
 [0.42857143 0.5       ]]
[[0.2        0.33333333]
 [0.42857143 0.5       ]]
[[1.         1.41421356]
 [1.73205081 2.        ]]


Numpy provides many useful functions for performing computations on arrays; one of the most useful is sum:

In [25]:
import numpy as np

x = np.array([[1,2],[3,4]])

print(np.sum(x))  # Compute sum of all elements; prints "10"
print(np.sum(x, axis=0))  # Compute sum of each column; prints "[4 6]"
print(np.sum(x, axis=1))  # Compute sum of each row; prints "[3 7]"

10
[4 6]
[3 7]


# Exercise 2:  Explore the Digits - Data

Load the digits data using the ```load_digits()``` function provided by scikit-learn. Explore the data and try to answer the following questions:
1. What is the number of **samples** and what is the **dimension** of the data?
2. What is the **data-type** (categorical, numerical, ...)?
3. Is the dataset suited for a **supervised** algorithm?
4. Is the dataset suited for an **unsupervised** algorithm?

*Hints:*
* Each dataset comes with a description that can be viewed using ```print(data.DESCR)```.
* ```type(x)``` gives us the type of an object in python.
* ```x.shape``` tells us the shape (dimensions in each axis) of a ```numpy``` array. 


In [26]:
import sklearn
import matplotlib.pyplot as plt
from sklearn.datasets import load_digits

digits = load_digits()
print(digits.data)
print('Nr. of Dimensions ', digits.data.ndim)
print('Shape: ', digits.data.shape)

[[ 0.  0.  5. ...  0.  0.  0.]
 [ 0.  0.  0. ... 10.  0.  0.]
 [ 0.  0.  0. ... 16.  9.  0.]
 ...
 [ 0.  0.  1. ...  6.  0.  0.]
 [ 0.  0.  2. ... 12.  0.  0.]
 [ 0.  0. 10. ... 12.  1.  0.]]
.. _digits_dataset:

Optical recognition of handwritten digits dataset
--------------------------------------------------

**Data Set Characteristics:**

    :Number of Instances: 1797
    :Number of Attributes: 64
    :Attribute Information: 8x8 image of integer pixels in the range 0..16.
    :Missing Attribute Values: None
    :Creator: E. Alpaydin (alpaydin '@' boun.edu.tr)
    :Date: July; 1998

This is a copy of the test set of the UCI ML hand-written digits datasets
https://archive.ics.uci.edu/ml/datasets/Optical+Recognition+of+Handwritten+Digits

The data set contains images of hand-written digits: 10 classes where
each class refers to a digit.

Preprocessing programs made available by NIST were used to extract
normalized bitmaps of handwritten digits from a preprinted form. From a
total of

1) 1797 Samples, 64 Dimensions

2) Categorical data, as it contains 10 classes (labels)

3) Yes, as we can give feedback to the algorithm by providing not only the images but also the lables (Typical classification problem)


4) No

# Exercise 3:  Explore the Diabetes - Data

Load the digits data using the ```load_digits()``` function provided by scikit-learn. Explore the data and try to answer the following questions:
1. What is the number of **samples** and what is the **dimension** of the data?
2. What is the **data-type** (categorical, numerical, ...)?
3. Is the dataset suited for a **supervised** algorithm?
4. Is the dataset suited for an **unsupervised** algorithm?

*Hints:*
* Each dataset comes with a description that can be viewed using ```print(data.DESCR)```.
* ```type(x)``` gives us the type of an object in python.
* ```x.shape``` tells us the shape (dimensions in each axis) of a ```numpy``` array. 

In [27]:
import sklearn
import matplotlib.pyplot as plt
from sklearn.datasets import load_diabetes

diabetes = load_diabetes()
print(diabetes.data[:2,:])
print('Nr. of Dimensions ', diabetes.data.ndim)
print('Shape: ', diabetes.data.shape)
print(diabetes.target)

[[ 0.03807591  0.05068012  0.06169621  0.02187239 -0.0442235  -0.03482076
  -0.04340085 -0.00259226  0.01990749 -0.01764613]
 [-0.00188202 -0.04464164 -0.05147406 -0.02632753 -0.00844872 -0.01916334
   0.07441156 -0.03949338 -0.06833155 -0.09220405]]
.. _diabetes_dataset:

Diabetes dataset
----------------

Ten baseline variables, age, sex, body mass index, average blood
pressure, and six blood serum measurements were obtained for each of n =
442 diabetes patients, as well as the response of interest, a
quantitative measure of disease progression one year after baseline.

**Data Set Characteristics:**

  :Number of Instances: 442

  :Number of Attributes: First 10 columns are numeric predictive values

  :Target: Column 11 is a quantitative measure of disease progression one year after baseline

  :Attribute Information:
      - age     age in years
      - sex
      - bmi     body mass index
      - bp      average blood pressure
      - s1      tc, total serum cholesterol
      - s2 

1) 442 Samples, 10 Dimensions

2) Numerical

3) Yes, a Regression can be applied

4) No